src/main/java/org/apache/flink/benchmark/ProcessingTimerBenchmark.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.benchmark;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
+import org.apache.flink.streaming.api.functions.sink.DiscardingSink;
+import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;
+import org.apache.flink.util.Collector;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.OperationsPerInvocation;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+
+public class ProcessingTimerBenchmark extends BenchmarkBase {
+
+    private static final int PARALLELISM = 4;
+    private static final int PROCESSING_TIMERS_PER_INVOCATION = 4_000;
+    private static final int PROCESSING_TIMERS_PER_RECORD = 4_000;
+
+    private static CountDownLatch LATCH = null;
+
+    public static void main(String[] args)
+            throws RunnerException {
+        Options options = new OptionsBuilder()
+                .verbosity(VerboseMode.NORMAL)
+                .include(".*" + ProcessingTimerBenchmark.class.getCanonicalName() + ".*")
+                .build();
+
+        new Runner(options).run();
+    }
+
+    @Benchmark
+    @OperationsPerInvocation(value = ProcessingTimerBenchmark.PROCESSING_TIMERS_PER_INVOCATION)
+    public void fireProcessingTimers(FlinkEnvironmentContext context) throws Exception {
+        LATCH = new CountDownLatch(PROCESSING_TIMERS_PER_INVOCATION);
```

Comment:
smaller issue: with parallelism of 1, I would replace this with single shot latch and a non-volatile counter. With just a single operator firing timers, that single operator wouldn't need to synchronise with anyone to decide whether it's time to trigger the latch or not and it would avoid synchronisation overheads for the count down.